### PR TITLE
Improve build time

### PR DESCRIPTION
**Pull request by vasil-pashov**
- Fix compilation errors with clang-18 due to Werror
- Run include-what-you-use to reduce includes and add forward declarations

**TLDR**
Using the suggestions from [include-what-you-use](https://github.com/include-what-you-use/include-what-you-use) makes full builds ~10% faster sparing 20s on a local machine (Processor: 13th Gen Intel(R) Core(TM) i7-1370P, 1900 Mhz, 14 Core(s), 20 Logical Processor(s)) building with 20 threads.

Build info:
OS: Ubuntu (WSL)
Compiler: Clang 18
Linker: Mold
Build system: Ninja
(Clang and Mold are custom builds using additional flags -O3 -march=native -flto=full, arcticdb is built as usual)
| #      | Master (debug) in seconds  | IWYU (debug) in seconds  |
|--------|-------------|-------------|
| 1      |      212.31 |      203.37 |
| 2      |      234.02 |      212.33 |
| 3      |      234.24 |      213.05 |
| 4      |      233.63 |      213.32 |
| 5      |      233.27 |      213.76 |
| 6      |      233.79 |      213.07 |
| 7      |      234.13 |      213.23 |
| 8      |      232.67 |      212.27 |
| 9      |      233.18 |      212.25 |
| 10     |      233.87 |      212.58 |
| 11     |      233.75 |      213.21 |
| 12     |      233.63 |      213.16 |
| 13     |      232.89 |      213.87 |
| 14     |      233.64 |      213.39 |
| 15     |      233.12 |      213.54 |
| Mean   | 232.1426667 | 212.4266667 |
| Median |      233.63 |      213.16 |
| SD     | 5.505148845 | 2.557028595 |

#### Checklist

<details>
  <summary>
   Checklist for code changes...
  </summary>
 
 - [ ] Have you updated the relevant docstrings, documentation and copyright notice?
 - [ ] Is this contribution tested against [all ArcticDB's features](../docs/mkdocs/docs/technical/contributing.md)?
 - [ ] Do all exceptions introduced raise appropriate [error messages](https://docs.arcticdb.io/error_messages/)?
 - [ ] Are API changes highlighted in the PR description?
 - [ ] Is the PR labelled as enhancement or bug so it appears in autogenerated release notes?
</details>

### CODE REVIEW

File: cpp/arcticdb/codec/segment.hpp [hunk 1]
```diff
@@ -7,12 +7,13 @@
 
 #pragma once
 
+//#include <google/protobuf/io/zero_copy_stream_impl.h>  // lines 13-13
+//#include <codec/encoding_sizes.hpp>  // lines 12-12
+//#include <util/buffer_pool.hpp>  // lines 15-15
```
Comment (IvoDD) [inline cpp/arcticdb/codec/segment.hpp:12]: Maybe just remove these instead of commenting them out?

File: cpp/arcticdb/processing/aggregation.cpp [hunk 2]
```diff
@@ -72,14 +73,6 @@ namespace
         }
     }
 
-    inline util::BitMagic::enumerator::value_type deref(util::BitMagic::enumerator iter) {
-        return *iter;
-    }
-
-    inline std::size_t deref(std::size_t index) {
```
Comment (IvoDD) [inline cpp/arcticdb/processing/aggregation.cpp:79]: Maybe deleting unused functions is not always ideal? If they could be useful in the future it would be unfortunate to have to rewrite them.

Doesn't matter so much for these small functions, so I don't have a strong opinion on this.

File: cpp/arcticdb/storage/test/test_storage_factory.cpp [hunk 3]
```diff
@@ -1,73 +0,0 @@
...
-
-#include <unordered_map>
-#include <vector>
-#include <type_traits>
-
-namespace ac = arcticdb;
-namespace as = arcticdb::storage;
-namespace asl = arcticdb::storage::lmdb;
-
-TEST(TestStorageFactory, LmdbLookup) {
```
Comment (IvoDD) [inline cpp/arcticdb/storage/test/test_storage_factory.cpp:27]: Deleting tests doesn't look good. Are they moved somewhere? Or this file was forgotten in the CMake?

File: cpp/arcticdb/util/test/gtest_main.cpp [hunk 4]
```diff
@@ -7,7 +7,7 @@
 
 #include <gtest/gtest.h>
 #include <arcticdb/util/global_lifetimes.hpp>
-#include <pybind11/pybind11.h>  // Must not directly include Python.h on Windows
+#include <pybind11/pybind11.h>
```
Comment (IvoDD) [inline cpp/arcticdb/util/test/gtest_main.cpp:10]: Nit: A potentially useful comment got deleted. Fine to leave as is if comment is not useful.

**Review by IvoDD — APPROVED** [review]
